Source/DafnyCore/Resolver/PreType/UnderspecificationDetector.cs

```diff
@@ -0,0 +1,584 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.Diagnostics.Contracts;
+using Microsoft.Boogie;
+using Bpl = Microsoft.Boogie;
+
+namespace Microsoft.Dafny {
+  public class UnderspecificationDetector : ResolverPass {
+    public UnderspecificationDetector(Resolver resolver)
+      : base(resolver) {
+    }
+
+    /// <summary>
+    /// This method
+    ///     * checks that type inference was able to determine all types
+    ///     * checks that shared destructors in datatypes are in agreement
+    ///     * checks that ORDINAL is used only where allowed
```

Comment:
Most of these checks seems to be implemented as methods in the class `DetectUnderspecificationVisitor`. Might be good to move / add comments closer to the definitions.

Reply:
I added further comments (here) that explain, more specifically, where those checks are performed. In doing so, I noticed that the last thing on the list isn't performed here at all (it's done in Pass 0 of resolution, but well after types have been inferred), so I removed that comment.